tls/s2n_internal.h

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#pragma once
+
+#include "api/s2n.h"
+
+/*
+ * Internal APIs used for exploring experimental APIs.
+ *
+ * These APIs change the behavior of S2N in potentially dangerous ways and should only be
+ * used for testing purposes. All Internal APIs are subject to change without notice.
+ */
+
+
+/*
+ * Return a pointer to the config set on the connection.
+ *
+ * This function will return a pointer to the config set by `s2n_connection_set_config`.
+ * It will return NULL prior to `s2n_connection_set_config` being called and a config
+ * being set by the application.
+ *
+ * Caution: A config can be associated with multiple connections and should not be
+ * modified after it has been built. Doing so is undefined behavior.
+ *
+ * WARNING: this method is still considered experimental and will not always report
+ * the correct alert description. It may be used for testing and logging, but
+ * not relied on for production logic.
+ */
+
+S2N_API
+extern struct s2n_config *s2n_internal_connection_get_config(struct s2n_connection *conn);
```

Comment:
I'm not sure about including "internal" in the name :\

Reply:
The rust bindings generator does a match on `s2n_internal.*` at the moment. When I tried to include `s2n_connection` it included all the `s2n_connection` functions.

I wonder if there is a way around this restriction.. lemme take a look.

If we dont want to include `s2n_internal` then we would have to specify the exact name, which is kinda painful. Is there a more succinct prefix that might work?

Reply:
How do we pick up all the methods in s2n.h, regardless of prefix?

Reply:
> How do we pick up all the methods in s2n.h, regardless of prefix?

good point.. let me investigate

Reply:
i would prefer to keep the internal for two reasons:
- it makes bindgen bindings generation easier and less hacky
- like cameron said, calling it internal makes it easier for readers to tell what it is

Reply:
discussed offline and going with no internal prefix